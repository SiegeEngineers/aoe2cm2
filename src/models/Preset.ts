import Turn from "./Turn";
import {CivilisationEncoder} from "../util/CivilisationEncoder";
import {Assert} from "../util/Assert";
import Civilisation from "./Civilisation";
import DraftOption from "./DraftOption";
import {Util} from "../util/Util";

class Preset {

    public static readonly EMPTY: Preset = new Preset('', [], []);

    public static readonly NEW: Preset = new Preset('', Civilisation.ALL, []);

    public static readonly SAMPLE: Preset = new Preset('Default Preset', Civilisation.ALL, [
        Turn.HOST_GLOBAL_BAN,
        Turn.GUEST_GLOBAL_BAN,
        Turn.HOST_HIDDEN_BAN,
        Turn.GUEST_HIDDEN_BAN,
        Turn.REVEAL_ALL,
        Turn.HOST_GLOBAL_PICK,
        Turn.GUEST_GLOBAL_PICK,
        Turn.HOST_PICK,
        Turn.GUEST_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.HOST_HIDDEN_PICK,
        Turn.GUEST_HIDDEN_PICK,
        Turn.REVEAL_ALL,
        Turn.HOST_HIDDEN_SNIPE,
        Turn.GUEST_HIDDEN_SNIPE,
        Turn.REVEAL_ALL
    ]);

    public static readonly SIMPLE: Preset = new Preset('Simple Preset', Civilisation.ALL, [
        Turn.HOST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_BAN,
        Turn.GUEST_NONEXCLUSIVE_PICK,
        Turn.HOST_NONEXCLUSIVE_PICK
    ]);

    public readonly name: string;
    public presetId?: string
    public readonly encodedCivilisations?: string;
    public readonly draftOptions?: DraftOption[];
    public readonly turns: Turn[];

    constructor(name: string, draftOptions: DraftOption[], turns: Turn[] = [], presetId?: string) {
        this.name = name;
        this.presetId = presetId;
        if (Util.isCivilisationArray(draftOptions)) {
            this.encodedCivilisations = CivilisationEncoder.encodeCivilisationArray(draftOptions);
        } else {
            this.draftOptions = draftOptions;
        }
        this.turns = turns;
    }

    public static fromPojo(preset: { name: string, encodedCivilisations?: string, draftOptions?: DraftOption[], turns: Turn[], presetId?: string } | undefined): Preset | undefined {
        if (preset === undefined) {
            return undefined;
        }
        Assert.isString(preset.name);
        Assert.isOptionalString(preset.encodedCivilisations);
        Assert.isOptionalString(preset.presetId);
        let draftOptions: DraftOption[] = [];
        if (preset.encodedCivilisations) {
            draftOptions = CivilisationEncoder.decodeCivilisationArray(preset.encodedCivilisations);
        } else if (preset.draftOptions) {
            draftOptions = preset.draftOptions;
        }
        return new Preset(preset.name, draftOptions, Turn.fromPojoArray(preset.turns), preset.presetId);
    }

    public addTurn(turn: Turn) {
        this.turns.push(turn);
    }

    get options(): DraftOption[] {
        if (this.encodedCivilisations) {
            return CivilisationEncoder.decodeCivilisationArray(this.encodedCivilisations);
        }
        if (this.draftOptions) {
            return this.draftOptions;
        }
        throw new Error('Invalid Preset without either encodedCivilisations or draftOptions');
    }
}

export default Preset;