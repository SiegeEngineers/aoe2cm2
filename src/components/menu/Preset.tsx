import * as React from "react";
import {PresetUtil} from "../../util/PresetUtil";
import {default as ModelPreset} from "../../models/Preset"
import TurnRow from "../draft/TurnRow";
import NewDraftButton from "../NewDraftButton";
import CustomisePresetButton from "../../containers/CustomisePresetButton";
import CopyableInput from "../draft/CopyableInput";
import {PresetOptionCheckbox} from "../PresetEditor/PresetOptionCheckbox";
import Civilisation from "../../models/Civilisation";
import DraftOptionPanelType from "../../constants/DraftOptionPanelType";
import DraftOptionPanel from "../draft/DraftOptionPanel";
import DraftOption from "../../models/DraftOption";
import {ApplicationState} from "../../types";
import {Dispatch} from "redux";
import * as actions from "../../actions";
import {connect} from "react-redux";

interface IState {
    preset?: ModelPreset;
    presetExists: boolean;
}

interface IProps {
    iconStyle: string;
}

class Preset extends React.Component<IProps, IState> {

    state: IState = {presetExists: true};

    componentDidMount() {
        const presetId = PresetUtil.getIdFromUrl();
        if (presetId !== undefined) {
            this.loadPreset(presetId);
        }
    }

    public render() {
        if (!this.state.presetExists) {
            return (
                <div className="content box">
                    <p>Could not find this preset, sorry :-(</p>
                </div>
            );
        }
        if (this.state.preset !== undefined) {

            const presetCivilisations = this.state.preset.options;
            let civs;
            let itemAlignment = '';
            if (this.state.preset.encodedCivilisations) {
                civs = Civilisation.ALL.map((value: Civilisation, index: number) =>
                    <PresetOptionCheckbox presetOptions={presetCivilisations} value={value}
                                          key={index}
                                          disabled={true}/>)
            } else {
                civs = presetCivilisations.map((value: DraftOption, index: number) =>
                    <DraftOptionPanel draftOption={value} active={false}
                                      draftOptionPanelType={DraftOptionPanelType.CHOICE} nextAction={0}
                                      displayOnly={true} iconStyle={this.props.iconStyle}/>);
                itemAlignment = ' flex-justify-center';
            }

            return (
                <div className='content box'>
                    <h3 className="has-text-centered">{this.state.preset.name}</h3>

                    <TurnRow turns={this.state.preset.turns}/>

                    <div className={"is-flex" + itemAlignment} style={{flexDirection: 'row', flexWrap: 'wrap'}}>
                        {civs}
                    </div>

                    <div className="columns is-mobile mt-4">
                        <div className="column is-7 buttons">
                            <br/>
                            <NewDraftButton preset={this.state.preset}/>
                            <CustomisePresetButton preset={this.state.preset}/>
                        </div>
                        <div className="column is-5">
                            <CopyableInput content={window.location.href} before={'preset.shareThisPreset'}
                                           length={35}/>
                        </div>
                    </div>
                </div>
            );
        }
        return (
            <div className="content box">
                <p>Loading…</p>
            </div>
        );
    }


    private loadPreset = (presetId: string) => {
        const request = new XMLHttpRequest();
        request.open('GET', '/api/preset/' + presetId, true);
        request.onreadystatechange = () => {
            if (request.readyState === XMLHttpRequest.DONE && request.status === 200) {
                const result = JSON.parse(request.responseText);
                const preset = ModelPreset.fromPojo(result);
                this.setState({preset});
                document.title = `Preset "${preset?.name}" – AoE2 Captains Mode`;
            } else if (request.readyState === XMLHttpRequest.DONE && request.status === 404) {
                this.setState({presetExists: false});
                document.title = 'Preset not found – AoE2 Captains Mode';
            }
        };
        request.send();
    };
}


function mapStateToProps(state: ApplicationState) {
    return {
        iconStyle: state.iconStyle.iconStyle,
    };
}

function mapDispatchToProps(dispatch: Dispatch<actions.Action>) {
    return {};
}

export default connect(mapStateToProps, mapDispatchToProps)(Preset);
