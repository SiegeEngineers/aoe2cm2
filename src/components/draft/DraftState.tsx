import * as React from 'react';
import {default as ModelPlayer} from '../../constants/Player';
import PlayerDraftState from "../../containers/PlayerDraftState";
import Preset from "../../models/Preset";

interface IProps {
    nameHost: string;
    nameGuest: string;
    preset: Preset;
    flipped: boolean;
    smooch: boolean;
    simplifiedUI: boolean;
}

class DraftState extends React.Component<IProps, object> {
    public render() {
        return (
            <div className="columns is-mobile player-panels">
                <PlayerDraftState preset={this.props.preset}
                                  player={ModelPlayer.HOST}
                                  name={this.props.nameHost}
                                  key={ModelPlayer.HOST}
                                  flipped={this.props.flipped}
                                  smooch={this.props.smooch}
                                  simplifiedUI={this.props.simplifiedUI}/>
                <PlayerDraftState preset={this.props.preset}
                                  player={ModelPlayer.GUEST}
                                  name={this.props.nameGuest}
                                  key={ModelPlayer.GUEST}
                                  flipped={this.props.flipped}
                                  smooch={this.props.smooch}
                                  simplifiedUI={this.props.simplifiedUI}/>
            </div>
        );
    }
}

export default DraftState;
