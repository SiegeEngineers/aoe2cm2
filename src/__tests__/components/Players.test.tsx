import {shallow} from "enzyme";
import * as React from "react";
import DraftState from "../../components/draft/DraftState";
import Preset from "../../models/Preset";


it('DraftOptionGrid renders correctly', () => {
    const component = shallow(<DraftState nameHost={'Sneaky Saladin'} nameGuest={'Beastly Barbarossa'}
                                          preset={Preset.SAMPLE}/>);
    expect(component).toMatchSnapshot();
});
