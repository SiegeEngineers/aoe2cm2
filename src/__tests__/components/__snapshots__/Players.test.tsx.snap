// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DraftOptionGrid renders correctly 1`] = `
<Fragment>
  <div
    className="columns is-mobile player-panels"
  >
    <Connect(Component)
      key="HOST"
      name="Sneaky Saladin"
      player="HOST"
      preset={
        Preset {
          "categoryLimits": Object {
            "ban": Object {},
            "pick": Object {},
          },
          "draftOptions": undefined,
          "encodedCivilisations": "3e3fffffffefff",
          "name": "Default Preset",
          "presetId": undefined,
          "turns": Array [
            Turn {
              "action": "BAN",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "HOST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "BAN",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "GUEST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "BAN",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "BAN",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "REVEAL_ALL",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "NONE",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "NONE",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "HOST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "GUEST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "EXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "EXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "REVEAL_ALL",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "NONE",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "NONE",
            },
            Turn {
              "action": "SNIPE",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "SNIPE",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "REVEAL_ALL",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "NONE",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "NONE",
            },
          ],
        }
      }
    />
    <Connect(Component)
      key="GUEST"
      name="Beastly Barbarossa"
      player="GUEST"
      preset={
        Preset {
          "categoryLimits": Object {
            "ban": Object {},
            "pick": Object {},
          },
          "draftOptions": undefined,
          "encodedCivilisations": "3e3fffffffefff",
          "name": "Default Preset",
          "presetId": undefined,
          "turns": Array [
            Turn {
              "action": "BAN",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "HOST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "BAN",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "GUEST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "BAN",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "BAN",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "REVEAL_ALL",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "NONE",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "NONE",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "HOST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "GUEST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "EXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "EXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "REVEAL_ALL",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "NONE",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "NONE",
            },
            Turn {
              "action": "SNIPE",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "SNIPE",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "REVEAL_ALL",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "NONE",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "NONE",
            },
          ],
        }
      }
    />
  </div>
  <div
    className="columns is-mobile player-panels"
  >
    <Connect(Component)
      key="Admin"
      name="Admin"
      player="NONE"
      preset={
        Preset {
          "categoryLimits": Object {
            "ban": Object {},
            "pick": Object {},
          },
          "draftOptions": undefined,
          "encodedCivilisations": "3e3fffffffefff",
          "name": "Default Preset",
          "presetId": undefined,
          "turns": Array [
            Turn {
              "action": "BAN",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "HOST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "BAN",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "GUEST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "BAN",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "BAN",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "REVEAL_ALL",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "NONE",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "NONE",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "HOST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "GUEST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "EXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "EXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "PICK",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "REVEAL_ALL",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "NONE",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "NONE",
            },
            Turn {
              "action": "SNIPE",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "HOST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "HOST",
            },
            Turn {
              "action": "SNIPE",
              "categories": Array [
                "default",
              ],
              "exclusivity": "NONEXCLUSIVE",
              "executingPlayer": "GUEST",
              "hidden": true,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "GUEST",
            },
            Turn {
              "action": "REVEAL_ALL",
              "categories": Array [
                "default",
              ],
              "exclusivity": "GLOBAL",
              "executingPlayer": "NONE",
              "hidden": false,
              "id": "mocked-uuid",
              "parallel": false,
              "player": "NONE",
            },
          ],
        }
      }
    />
  </div>
</Fragment>
`;
